CMakeLists.txt

```diff
@@ -2,156 +2,243 @@
+target_include_directories(pybind11 SYSTEM INTERFACE $<BUILD_INTERFACE:${PYTHON_INCLUDE_DIRS}>)
+
+if(CMAKE_VERSION VERSION_LESS 3.13)
+  target_compile_features(
+    pybind11
+    INTERFACE cxx_auto_type
+              cxx_constexpr
+              cxx_defaulted_functions
+              cxx_lambdas
+              cxx_nullptr
+              cxx_range_for
+              cxx_right_angle_brackets
+              cxx_strong_enums
+              cxx_user_literals
+              cxx_variadic_templates)
+else()
+  # This was added in CMake 3.8, but we are keeping a consistent breaking point for the config file
+  # at 3.13. A config generated by CMake 3.13+ can ony be read in 3.13+ due to the SHELL usage
+  # later, so this is safe to do.
+  target_compile_features(pybind11 INTERFACE cxx_std_11)
 endif()
 
-if(NOT (CMAKE_VERSION VERSION_LESS 3.0))  # CMake >= 3.0
-  # Build an interface library target:
-  add_library(pybind11 INTERFACE)
-  add_library(pybind11::pybind11 ALIAS pybind11)  # to match exported target
-  target_include_directories(pybind11 INTERFACE $<BUILD_INTERFACE:${PYBIND11_INCLUDE_DIR}>
-                                                $<BUILD_INTERFACE:${PYTHON_INCLUDE_DIRS}>
-                                                $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
-  target_compile_options(pybind11 INTERFACE $<BUILD_INTERFACE:${PYBIND11_CPP_STANDARD}>)
-
-  add_library(module INTERFACE)
-  add_library(pybind11::module ALIAS module)
-  if(NOT MSVC)
-    target_compile_options(module INTERFACE -fvisibility=hidden)
-  endif()
-  target_link_libraries(module INTERFACE pybind11::pybind11)
-  if(WIN32 OR CYGWIN)
-    target_link_libraries(module INTERFACE $<BUILD_INTERFACE:${PYTHON_LIBRARIES}>)
-  elseif(APPLE)
-    target_link_libraries(module INTERFACE "-undefined dynamic_lookup")
-  endif()
+add_library(module INTERFACE)
+add_library(pybind11::module ALIAS module)
+
+target_link_libraries(module INTERFACE pybind11::pybind11)
 
-  add_library(embed INTERFACE)
-  add_library(pybind11::embed ALIAS embed)
-  target_link_libraries(embed INTERFACE pybind11::pybind11 $<BUILD_INTERFACE:${PYTHON_LIBRARIES}>)
+# See https://github.com/Kitware/CMake/blob/master/Modules/CMakePlatformId.h.in for platform IDs
+# Note: CMake 3.15 allows $<PLATFORM_ID:Windows,Cygwin>
+target_link_libraries(
+  module
+  INTERFACE
+    "$<$<OR:$<PLATFORM_ID:Windows>,$<PLATFORM_ID:Cygwin>>:$<BUILD_INTERFACE:${PYTHON_LIBRARIES}>>")
+
+if(CMAKE_VERSION VERSION_LESS 3.13)
+  target_link_libraries(module INTERFACE "$<$<PLATFORM_ID:Darwin>:-undefined dynamic_lookup>")
+else()
+  # SHELL (3.12+) forces this to remain together, and link_options was added in 3.13+
+  target_link_options(module INTERFACE "$<$<PLATFORM_ID:Darwin>:SHELL:-undefined dynamic_lookup>")
 endif()
 
-if (PYBIND11_INSTALL)
+message(STATUS "CXX Compiler version: ${CMAKE_CXX_COMPILER_VERSION}")
+
+# Workaround for Python 2.7 and C++17 (C++14 as a warning) incompatibility
+set(clang_4plus
```

Comment:
Generator expressions are not the most readable representation, and I had a hard time figuring the end result here. Can you provide a human-readable sumary that goes beyond the `# Workaround ...`?

Reply:
Sure.